stormpath/src/main/java/com/stormpath/sdk/StormpathLogger.java

```diff
@@ -0,0 +1,127 @@
+package com.stormpath.sdk;
+
+import android.support.annotation.IntDef;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public abstract class StormpathLogger {
+
+    public static final int SILENT = 100;
+
+    /**
+     * Logging levels -- values match up with android.util.Log
+     */
+    public static final int VERBOSE = 2;
+
+    public static final int DEBUG = 3;
+
+    public static final int INFO = 4;
+
+    public static final int WARN = 5;
+
+    public static final int ERROR = 6;
+
+    public static final int ASSERT = 7;
+
+    @IntDef(value = {SILENT, VERBOSE, DEBUG, INFO, WARN, ERROR, ASSERT})
+    @interface LogLevel {
+
+    }
+
+    /**
+     * By default we don't log anything.
+     */
+    private int logLevel = SILENT;
+
+    public void setLogLevel(@LogLevel int logLevel) {
+        this.logLevel = logLevel;
+    }
+
+    protected abstract void log(@LogLevel int logLevel, String message);
+
+    protected void log(@LogLevel int logLevel, Throwable throwable, String message, Object... args) {
+        if (!isLoggable(logLevel)) {
+            return;
+        }
+
+        if (message == null) {
+            if (throwable == null) {
+                return; // Swallow message if it's null and there's no throwable.
+            }
+
+            message = getStackTraceString(throwable);
+        } else {
+            if (args.length > 0) {
+                message = String.format(message, args);
+            }
+            if (throwable != null) {
+                message += "\n" + getStackTraceString(throwable);
+            }
+        }
+
+        log(logLevel, message);
+    }
+
+    private boolean isLoggable(@LogLevel int level) {
+        return level <= this.logLevel;
+    }
+
+    private String getStackTraceString(Throwable t) {
+        // Don't replace this with Log.getStackTraceString() - it hides
+        // UnknownHostException, which is not what we want.
+        StringWriter sw = new StringWriter(256);
+        PrintWriter pw = new PrintWriter(sw, false);
+        t.printStackTrace(pw);
+        pw.flush();
+        return sw.toString();
+    }
+
```

Comment:
Added in https://github.com/stormpath/stormpath-sdk-android/commit/e34b9f701a5691be0cac5794345138e8ef92726f.